js/angular-crm-ui.js

```diff
@@ -130,6 +132,27 @@
       };
     })
 
+    // standalone: <a crm-ui-help="field_name">
+    // within crmUiField: <div crm-ui-field crm-title="My Field" crm-help="field_name">
+    .directive('crmUiHelp', function() {
+      return {
+        restrict: 'EA',
+        scope: {
+          crmUiHelp: '='
+        },
+        link: function (scope, element, attrs) {
+          element
+            .addClass('helpicon')
+            .attr('title', ts('%1 Help', {1: scope.crmUiHelp.title}))
+            .attr('href', '#')
+            .on('click', function(e) {
+              e.preventDefault();
+              CRM.help(scope.crmUiHelp.title, {id: scope.crmUiHelp.id, file: scope.$parent.$parent.$parent.helpFile});
```

Comment:
I agree that the triple-$parent is pretty smelly.

To cover our bases... this element is using *isolate scope*, which is a concept that can apply to directives (but not to basic controllers like "EditMailingCtrl" or top-level views like "edit.html"). The links seem geared toward the application developer who works with the controller and top-level-view. Both links give a brief mention to the subject:

```
-- Link #1
The one exception: Some directives can optionally create
an isolate scope and do not inherit from their parents.

-- Link #2
One exception to this rule is a directive that uses
scope: { ... } -- this creates an "isolate" scope that does
not prototypically inherit.
```

Isolate scopes don't inherit the parent's scope -- but you can still access the parent using ```$parent```. Isolate scope is particularly interesting for something like ```crmUiTab``` or ```crmUiWizard``` (or, generally, for directives that use ```transclude:true```). The directive author (writing ```crmUiTab```) can freely define variables in his isolate scope which are used for the tab-bar/tab-frame; the application author (writing ```EditMailingCtrl``` + ```edit.html```) can freely define variables in his scopes (both outside the tabs and inside the body of the tab); and there won't be any naming conflicts.

Isolate-scope helps encapsulate data. It serves roughly the same purpose as ```private``` scoping in OOP (although the mechanics are entirely different). In an OOP model, the tab-bar/tab-frame might have some private properties for managing the tab state; similarly, in Angular, the tab-bar/tab-frame directive might have some properties in isolate-scope for managing the tab state.

Isolate-scope decreases friction from naming-conflicts but increases friction from data-passing -- i.e. when you want to pass data between the applications' scopes and the directives' scopes, you have to do so explicitly. Is the trade-off "worth it"? My vote is "yes" based on (1) a general philosophy that components should have explicit contracts and (2) the practical benefit that isolate-scoping puts all responsibility on one person (the directive author explicitly manages data-passing) rather than splitting responsibility among multiple people (the directive author and the application authors need to avoid naming-conflicts). But that's not a religious view; there may be times when isolate-scope is overkill.

In any event... triple-$parent seems very smelly; it's basically bypassing three-levels of isolation. Bypassing one-level (single-$parent) may be appropriate sometimes... but three suggests either (a) some components are too brittle and/or (b) some of the intermediate components are missing some data-passing code.